Improve IEx autocomplete to support navigating map atom keys

myronmarston opened the pull request:
Eventually we may want to take this a step further and support autocompletion of the variable names themselves but I don't have time today to look into that and that would probably be a separate PR.

I noticed that 1.4.0.rc-0 was just released...any chance this could be included in 1.4?  This is a feature I've long wanted and I'd be sad to see it pushed out months :(.

myronmarston commented:
The build failure appears to be unrelated to my changes.

josevalim commented:
Instead of passing the evaluator as argument, can we also use the autocomplete_server and have it implement a function called `evaluator`, [similar to how we have done in the `env_aliases` function](https://github.com/seomoz/elixir/blob/ef2d240154d4fdb2b6c23cbd0fa2cf84af1bcaa4/lib/iex/lib/iex/autocomplete.ex#L199-L201)?

josevalim commented on lib/iex/lib/iex/autocomplete.ex:
```diff
@@ -269,6 +289,14 @@ defmodule IEx.Autocomplete do
   end
 
+  defp match_map_fields(map, hint) do
+    map
```
A `for` comprehension will be cleaner and faster:

```elixir
for {key, _value} <- map,
    is_atom(key),
    key = to_string(key),
    String.starts_with?(key, hint),
    do: %{kind: :map_key, name: key, value_is_map: is_map(value)}
```

myronmarston replied:
I'm glad to change that because I like `for` comprehensions...but I'm curious about your statement that it's faster given my use of stream operations (so the map is only traversed once).  Why are comprehensions faster?  On my team others have expressed a preference for piping a collection through a bunch of `Stream` operations instead of using for comprehensions for this kind of situation, but if comprehensions are faster I'd like to understand that so I can explain it to my teammates :).

josevalim replied:
@myronmarston a for-comprehension is only traversing as well and it will be faster because it will skip all the machinery behind streams that make them lazy. So it is perfect for this use case (when you don't actually need laziness but want to avoid multiple traversals).

Btw, please continue asking questions, regardless of the reason behind them. :)

myronmarston replied:
> Btw, please continue asking questions, regardless of the reason behind them. :)

I definitely will. I've never been shy about asking about such things :).

josevalim commented on lib/iex/lib/iex/autocomplete.ex:
```diff
@@ -336,8 +372,42 @@ defmodule IEx.Autocomplete do
+  end
+
+  defp get_variable_value(ast_node, evaluator) do
+    binding = IEx.Evaluator.get_binding(evaluator)
```
We should ask the evaluator to perform the computation for us. Asking it to copy the whole binding can be very very very expensive.

myronmarston replied:
Makes sense.  Will change.

myronmarston replied:
Would you want me to pass the `ast_node` to the evaulator?  Or should we convert the `ast_node` into a an `Access`-style key path list here and pass that?

josevalim commented on lib/iex/lib/iex/autocomplete.ex:
```diff
@@ -336,8 +372,42 @@ defmodule IEx.Autocomplete do
+  defp get_variable_value(ast_node, evaluator) do
+    binding = IEx.Evaluator.get_binding(evaluator)
+
+    {_, value} = Macro.postwalk(ast_node, {:binding, binding}, fn
```
The binding is a keyword list. Is there any reason why we need Macro.postwalk? :)

myronmarston replied:
It's there to walk the AST when you are dealing with nested map access like here:

https://github.com/elixir-lang/elixir/pull/5488/files/ef2d240154d4fdb2b6c23cbd0fa2cf84af1bcaa4#diff-9f4689bfff76a705f5caf00680a347b6R134

The binding is just the starting accumulator.

Is there a better way to walk the nested map access AST to get the value out of it?

josevalim replied:
Hrm, that's too generate for what you are trying to build in the sense that I don't know if it will actually work or fail in other scenarios. My suggestion would be to convert `ast_node` to a list of variable accesses. And then use Enum.reduce. In pseudo code:

    {var, fields} = extract_var_and_fields(ast_node)
    value = IEx.Evaluator.binding(evaluator, var)
    Enum.reduce fields, {:ok, value}, fn
      field, {:ok, acc} when is_map(acc) -> Map.fetch(acc, field)
      _field, _acc -> :error
    end)

josevalim replied:
To answer your question above, we can either get one specific binding from the evaluator, or also pass it a list of fields and have it return only the final value. I believe having it receive the list of fields would be best, maybe something like: `IEx.Evaluator.field(evaluator, var, fields)`.

josevalim replied:
`var` is an atom and `fields` is a list of atoms.

myronmarston replied:
In `extract_var_and_fields`, how would I go about extracting `var` and `fields` w/o using `Most.postwalk`?  Seems like the AST has to be traversed and that's what it's for...

Or do you mean I should still use `Macro.postwalk` but use it only for extracting the variable name and fields and not traverse the binding as part of that?

josevalim replied:
I would try doing manual recursion. I believe, for example, your code would correctly postwalk on something like: `[foo, bar.baz, bar.bat]`, which is not what we want.

myronmarston commented:
> Instead of passing the evaluator as argument, can we also use the autocomplete_server and have it implement a function called evaluator, similar to how we have done in the env_aliases function?

Yes.  But I'd like to ask: why is that better?  In terms of referential transparency, it seems better to pass args to functions instead of making those function query something else to get the args they need.

josevalim commented:
@myronmarston 

1. I believe passing the argument is better

2. However, I don't believe a separate process should look at the internals (Process.info) of another server unless the server exposes this through an API

So ideally we would not rely on `Application.get_env` and pass it as argument. But I don't believe we should access the process dictionary of a separate process. What if we pass the server name as argument? Would you be up to changing env_aliases to receive the server name as well? What is neat is that we invoke only functions of zero-arity on the server, which means you can use a map during tests.

josevalim commented:
> I noticed that 1.4.0.rc-0 was just released...any chance this could be included in 1.4? This is a feature I've long wanted and I'd be sad to see it pushed out months :(.

If we do another release candidate we may include this change, otherwise I am afraid we should not merge it in without getting extensive enough testing.

myronmarston commented:
OK, @josevalim, I think I've addressed all your feedback except for this bit:

> What if we pass the server name as argument? Would you be up to changing env_aliases to receive the server name as well?

It wasn't clear to me what you wanted here so I didn't do this.  If you still want it, please clarify with some pseudo code.  Otherwise maybe what I have is sufficient or someone can do that in another PR.

> If we do another release candidate we may include this change, otherwise I am afraid we should not merge it in without getting extensive enough testing.

Makes sense.  I'd vote for another RC :).  I'd definitely run that RC locally so I could use this and test it out :).

On a completely unrelated note: I was surprised to see that GenStage wasn't included in 1.4.0.  Is it not quite stable enough to be included?  We are using it in production.

josevalim commented on lib/iex/lib/iex/evaluator.ex:
```diff
@@ -24,6 +24,21 @@ defmodule IEx.Evaluator do
+  """
+  @spec value_from_binding(pid, atom, [atom]) :: {:ok, any} | :error
+  def value_from_binding(evaluator, var_name, map_key_path) do
+    send evaluator, {:value_from_binding, self(), var_name, map_key_path}
```
The good practice here is to build a reference, send it to the server, and match on it on the response:

```elixir
ref = make_ref()
send evaluator, {:value_from_binding, ref, self(), var_name, map_key_path}

receive do
  {^ref, result} -> result
after
  5000 -> :error
end
```

This is for many reasons:

  1. If for some reason the evaluator cannot reply after 5000 miliseconds, it will eventually reply and a future request for a binding may return the value of a previous request (which will be sitting in the autocomplete process inbox)

  2. the VM can optimize that code snippet really well. It knows the time the reference was built and because they are unique, it inserts markers in the current process message queue to avoid traversing it from the beginning when looking back for the reply

myronmarston replied:
Thanks, I often use refs when messaging but forgot to do so here.  I'm glad, though; it gave you the opportunity to teach me some things about it that I didn't know, like the VM optimization :).

Anyhow, fixed.

josevalim commented on lib/iex/lib/iex/server.ex:
```diff
@@ -43,6 +43,23 @@ defmodule IEx.Server do
+  def evaluator() do
+    case IEx.Server.whereis() do
+      nil -> nil
+      server ->
```
Sorry, maybe I was not clear enough in my original request. It is OK to look into the process dictionary here* as long as the function that looks into the process dictionary is in the same module as the process running the server. So we can revert back to the process dictionary here since we have encapsulated it in the IEx.Server module.

*it is not really ok to use the process dictionary. the only reason we use here is to avoid deadlocks when the server, evaluator or autocomplete need to access each other while the server is executing a function on the evaluator or automcplete and vice-versa.

myronmarston replied:
Fixed.

josevalim commented on lib/iex/lib/iex/server.ex:
```diff
@@ -195,6 +216,9 @@ defmodule IEx.Server do
       {:peek_env, receiver} ->
         send evaluator, {:peek_env, receiver}
         wait_input(state, evaluator, evaluator_ref, input)
+      {:get_evaluator, receiver} ->
```
We can remove this clause as I mentioned above. :)

myronmarston replied:
Fixed.

josevalim commented:
> It wasn't clear to me what you wanted here so I didn't do this. If you still want it, please clarify with some pseudo code.

What I meant is that we could remove the calls to `Application.get_env(:iex, :autocomplete_server)` and pass the server name as argument:

    def expand(hint, server \\ IEx.Server)

And then you pass the server to the functions calling `Application.get_env(:iex, :autocomplete_server)`. I was basically agreeing with you: using dependency injection is better. But we need to be consistent and use dependency injection in both places using the server.

josevalim commented:
> On a completely unrelated note: I was surprised to see that GenStage wasn't included in 1.4.0. Is it not quite stable enough to be included? We are using it in production.

I have answered this one here: https://elixirforum.com/t/elixir-v1-4-0-rc-0-released/2627 :)

myronmarston commented:
OK, I think I've addressed all your feedback, @josevalim.  PTAL.

josevalim commented:
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

josevalim commented:
Also backported to v1.4. :)

myronmarston commented:
> Also backported to v1.4. :)

Thank you, that makes me happy :).

As I mentioned above, I think we can take this a step further and support autocompletion of variable names (basically, when you start typing a bare word it would use variable names and kernel/imported functions for auto-complete).  Is that something you'd like to see included in 1.4 as well?

josevalim commented:
> Is that something you'd like to see included in 1.4 as well?

It depends on the time we can get a patch in. But overall that's functionality we definitely would like to have. It would also be nice to handle custom imports too.

myronmarston commented:
> It depends on the time we can get a patch in.

I'll see if I can get to it in the next couple days.  If it's not included in 1.4, I'll definitely understand; personally the map key thing is the main thing I wanted :).

> It would also be nice to handle custom imports too.

What do you mean by "custom imports"?

josevalim commented:
> What do you mean by "custom imports"?

Today if I do `import List`, it won't autocomplete the `flatten` function.
